Source/WebGPU/WGSL/TypeCheck.cpp

```diff
@@ -124,10 +126,55 @@ class UnificationContext {
     }
 
     SetForScope<UnificationContext*> m_setForScope;
-    Vector<std::pair<Type*, Type*>> m_constraints;
-    Vector<std::pair<Type*, SubstituionCallback>> m_substitutions;
+    WTF::Vector<std::pair<Type*, Type*>> m_constraints;
+    WTF::Vector<std::pair<Type*, SubstituionCallback>> m_substitutions;
 };
 
+TypeChecker::TypeChecker(AST::ShaderModule& shaderModule)
+    : m_shaderModule(shaderModule)
+    , m_typeConstrutors(AST::ParameterizedTypeName::NumberOfBaseTypes)
+{
+    m_abstractInt = allocateType<Primitive>(Primitive::AbstractInt);
+    m_abstractFloat = allocateType<Primitive>(Primitive::AbstractFloat);
+    m_void = allocateType<Primitive>(Primitive::Void);
+    m_bool = allocateType<Primitive>(Primitive::Bool);
+    m_i32 = allocateType<Primitive>(Primitive::I32);
+    m_u32 = allocateType<Primitive>(Primitive::U32);
+    m_f32 = allocateType<Primitive>(Primitive::F32);
+
+    ContextProvider::introduceVariable(AST::Identifier::make("void"_s), m_void);
+    ContextProvider::introduceVariable(AST::Identifier::make("bool"_s), m_bool);
+    ContextProvider::introduceVariable(AST::Identifier::make("i32"_s), m_i32);
+    ContextProvider::introduceVariable(AST::Identifier::make("u32"_s), m_u32);
+    ContextProvider::introduceVariable(AST::Identifier::make("f32"_s), m_f32);
+
+#define FOR_EACH_BASE(f) \
+    f(Vec2, Vector, static_cast<uint8_t>(2)) \
+    f(Vec3, Vector, static_cast<uint8_t>(3)) \
+    f(Vec4, Vector, static_cast<uint8_t>(4)) \
+    f(Mat2x2, Matrix, static_cast<uint8_t>(2), static_cast<uint8_t>(2)) \
+    f(Mat2x3, Matrix, static_cast<uint8_t>(2), static_cast<uint8_t>(3)) \
+    f(Mat2x4, Matrix, static_cast<uint8_t>(2), static_cast<uint8_t>(4)) \
+    f(Mat3x2, Matrix, static_cast<uint8_t>(3), static_cast<uint8_t>(2)) \
+    f(Mat3x3, Matrix, static_cast<uint8_t>(3), static_cast<uint8_t>(3)) \
+    f(Mat3x4, Matrix, static_cast<uint8_t>(3), static_cast<uint8_t>(4)) \
+    f(Mat4x2, Matrix, static_cast<uint8_t>(4), static_cast<uint8_t>(2)) \
+    f(Mat4x3, Matrix, static_cast<uint8_t>(4), static_cast<uint8_t>(3)) \
+    f(Mat4x4, Matrix, static_cast<uint8_t>(4), static_cast<uint8_t>(4))
+
+#define DEFINE_TYPE_CONSTRUCTOR(base, type, ...) \
```

Comment:
Can't this be done with templates instead?

Reply:
Good point, seems like it should work! I'll refactor it when I move to Intrinsics.